staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -494,9 +495,16 @@ func (s sortableStoreElements) Swap(i, j int) {
 // with their ResourceVersion and the name of the index, if any, that was used.
 func (w *watchCache) WaitUntilFreshAndList(ctx context.Context, resourceVersion uint64, matchValues []storage.MatchValue) (result []interface{}, rv uint64, index string, err error) {
 	if utilfeature.DefaultFeatureGate.Enabled(features.ConsistentListFromCache) && w.notFresh(resourceVersion) {
-		w.waitingUntilFresh.Add()
-		err = w.waitUntilFreshAndBlock(ctx, resourceVersion)
-		w.waitingUntilFresh.Remove()
+		var requestWatchProgressSupported bool
+		requestWatchProgressSupported, err = etcdfeature.DefaultFeatureSupportChecker.Supports(storage.RequestWatchProgress)
+		if err != nil {
+			return result, rv, index, err
+		}
+		if requestWatchProgressSupported {
+			w.waitingUntilFresh.Add()
+			err = w.waitUntilFreshAndBlock(ctx, resourceVersion)
+			w.waitingUntilFresh.Remove()
+		}
```

Comment:
```suggestion
		} else {
			err = w.waitUntilFreshAndBlock(ctx, resourceVersion)
		}
```

Reply:
Updated.

Reply:
This look a bit odd to me. https://github.com/kubernetes/kubernetes/blob/12217672a3511f5e461ecb87b3e22422ab30bdbf/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go#L501C1-L502C1
Both `err = w.waitUntilFreshAndBlock(ctx, resourceVersion)` inside the `if` and the `else` looks the same.

Reply:
Please define a function
```
func (w *watchCache) shouldRegisterAsWaitingForFresh() (bool, error) {
	if !utilfeature.DefaultFeatureGate.Enabled(features.ConsistentListFromCache) || !w.notFresh(resourceVersion) {
		return false, nil
	}
	return etcdfeature.DefaultFeatureSupportChecker.Supports(storage.RequestWatchProgress)
}
```
And change the function to
```
func (w *watchCache) WaitUntilFreshAndList(ctx context.Context, resourceVersion uint64, matchValues []storage.MatchValue) (result []interface{}, rv uint64, index string, err error) {
	var shouldRegiste bool
	shouldRegister, err = w.shouldRegisterAsWaitingForFresh()
	if err != nil {
		return result, rv, index, err
	}
	if shouldRegister {
		w.waitingUntilFresh.Add()
		err = w.waitUntilFreshAndBlock(ctx, resourceVersion)
		w.waitingUntilFresh.Remove()
	} else {
		err = w.waitUntilFreshAndBlock(ctx, resourceVersion)
	}
...
```

Reply:
Should i change the usage of `etcdfeature.DefaultFeatureSupportChecker.Supports(storage.RequestWatchProgress)` in other parts too? to check `DefaultFeatureGate.Enabled(features.ConsistentListFromCache)` first or not?
https://github.com/kubernetes/kubernetes/blob/12217672a3511f5e461ecb87b3e22422ab30bdbf/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L724

Reply:
Don't think it's as useful. In the previous case we also checked `w.notFresh`